metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py

```diff
@@ -0,0 +1,525 @@
+    return update_needed
+
+
+class DataProduct(ConfigModel):
+    """This is a DataProduct class which represents a DataProduct
+
+    Args:
+        id (str): The id of the Data Product
+        domain (str): The domain that the Data Product belongs to. Either as a name or a fully-qualified urn.
+        owners (Optional[List[str, Ownership]]): A list of owners and their types.
+        display_name (Optional[str]): The name of the Data Product to display in the UI
+        description (Optional[str]): A documentation string for the Data Product
+        tags (Optional[List[str]]): An array of tags (either bare ids or urns) for the Data Product
+        terms (Optional[List[str]]): An array of terms (either bare ids or urns) for the Data Product
+        assets (List[str]): An array of entity urns that are part of the Data Product
+    """
+
+    class Ownership(ConfigModel):
+        id: str
+        type: str
+
+        @pydantic.validator("type")
+        def ownership_type_must_be_mappable(cls, v: str) -> str:
+            _ownership_types = [
+                OwnershipTypeClass.BUSINESS_OWNER,
+                OwnershipTypeClass.CONSUMER,
+                OwnershipTypeClass.DATA_STEWARD,
+                OwnershipTypeClass.DATAOWNER,
+                OwnershipTypeClass.DELEGATE,
+                OwnershipTypeClass.DEVELOPER,
+                OwnershipTypeClass.NONE,
+                OwnershipTypeClass.PRODUCER,
+                OwnershipTypeClass.STAKEHOLDER,
+                OwnershipTypeClass.TECHNICAL_OWNER,
+            ]
+            if v.upper() not in _ownership_types:
+                raise ValueError(f"Ownership type {v} not in {_ownership_types}")
+
+            return v.upper()
+
+    id: str
+    domain: str
+    _resolved_domain_urn: str
+    assets: List[str]
+    display_name: Optional[str] = None
+    owners: Optional[List[Union[str, DataProduct.Ownership]]] = None
+    description: Optional[str] = None
+    tags: Optional[List[str]] = None
+    terms: Optional[List[str]] = None
+    properties: Optional[Dict[str, str]] = None
+    external_url: Optional[str] = None
+
+    @pydantic.validator("assets", each_item=True)
+    def assets_must_be_urns(cls, v: str) -> str:
+        try:
+            Urn.create_from_string(v)
+        except Exception as e:
+            raise ValueError(f"asset {v} is not an urn", e)
+
+        return v
+
+    @property
+    def urn(self) -> str:
+        if self.id.startswith("urn:li:dataProduct:"):
+            return self.id
+        else:
+            return f"urn:li:dataProduct:{self.id}"
+
+    # If domain is an urn, we cache it in the private _resolved_domain_urn field
+    # Otherwise, we expect the caller to populate this by using an external DomainRegistry
+    @pydantic.validator("domain")
+    def cache_resolved_domain_urn_if_urn(cls, v: str) -> str:
+        if isinstance(v, str) and v.startswith("urn:li:domain:"):
+            cls._resolved_domain_urn = v
```

Comment:
domain would be an instance variable but it looks like we're caching it in a class variable here?